index.js

```diff
@@ -196,14 +196,28 @@
     return _funcPath(false, path, implementations);
   }
 
+  /* istanbul ignore next */
+  //  functionName :: Function -> Nullable String
+  var functionName = 'name' in function f() {} ?
+      function functionName(f) { return f.name; } :
+      function functionName(f) {
+        var match = /function (\w*)/.exec(f);
+        return match && match[1];
+      };
+
   //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
   function $(_name, dependencies, requirements) {
     function getBoundMethod(_name) {
       var name = 'fantasy-land/' + _name;
       return requirements[_name] === Constructor ?
         function(typeRep) {
-          return funcPath([name], typeRep) ||
-                 implPath([/function (\w*)/.exec(typeRep)[1], name]);
+          var f = funcPath([name], typeRep);
+          if (f == null && typeof typeRep === 'function') {
+            var typeIdent = functionName(typeRep);
+            /* istanbul ignore else */
+            if (typeIdent) f = implPath([typeIdent, name]);
```

Comment:
Let's use `typeIdent != null` as the condition.

The other option is to follow Irakli's suggestion by having `functionName` return `''` rather than `null` in the failure case. We could then remove the conditional logic above:

```javascript
f = implPath([functionName(typeRep), name]);
```

I like this idea now that we're using `Function#name` if available, since `(function() {}).name` and `(x => x).name` evaluate to `''` in modern environments.